### PR TITLE
Fire extinguishers and sprays only hit a tile once.

### PR DESCRIPTION
**Pull request by Contrabang**
## What Does This PR Do

Fixes fire extinguishers/sprays hitting a turf multiple times if they couldnt move further towards their target. Refactored quite a bit of code too.

## Why It's Good For The Game
Fixes some nasty stuff

## Images of changes

https://github.com/ParadiseSS13/Paradise/assets/91113370/40bbf672-044a-4a3a-8671-0886c055924f

## Testing

See above. Could extinguish myself too, and also use the chair-fire extinguisher system where it sends you flying

## Changelog
:cl:
fix: Fire extinguishers and sprays only hit a tile once.
/:cl:

### CODE REVIEW

**Comment by Daylight2** [comment]
Does firing the extinguisher diagonally still hit that diagonal tile with multiple sprays?

**Comment by Contrabang** [comment]
> Does firing the extinguisher diagonally still hit that diagonal tile with multiple sprays?

Yes, but thats (kind of) intended. This is just a fix for walls allowing it to hurt the same turf (and contents) multiple times

**Comment by 1080pCat** [comment]
SOUL REMOVAL!!!!!!!!!!!!

File: code/game/objects/items/weapons/extinguisher.dm [hunk 1]
```diff
@@ -98,90 +98,74 @@
...
-							var/mob/living/M = atm
-							M.ExtinguishMob()
-
-					if(W.loc == my_target) break
-					sleep(2)
-	else
+	if(safety)
 		return ..()
+	if(src.reagents.total_volume < 1)
+		to_chat(user, "<span class='danger'>\The [src] is empty.</span>")
```
Comment (DGamerL) [inline code/game/objects/items/weapons/extinguisher.dm:104]:
```suggestion
		to_chat(user, "<span class='danger'>[src] is empty.</span>")
```

File: code/game/objects/items/weapons/extinguisher.dm [hunk 2]
```diff
@@ -98,90 +98,74 @@
...
+		return
+
+	if(world.time < src.last_use + 2 SECONDS)
+		return
+
+	src.last_use = world.time
+
+	if(reagents.chem_temp > 300 || reagents.chem_temp < 280)
+		add_attack_logs(user, target, "Sprayed with superheated or cooled fire extinguisher at Temperature [reagents.chem_temp]K")
+	playsound(src.loc, 'sound/effects/extinguish.ogg', 75, 1, -3)
```
Comment (DGamerL) [inline code/game/objects/items/weapons/extinguisher.dm:114]:
```suggestion
	playsound(src.loc, 'sound/effects/extinguish.ogg', 75, TRUE, -3)
```

File: code/game/objects/effects/effect_system/effects_water.dm [hunk 3]
```diff
@@ -24,6 +24,23 @@
...
+			reagents.reaction(atm)
+			if(isliving(atm)) //For extinguishing mobs on fire
+				var/mob/living/M = atm
+				M.ExtinguishMob()
+
+		if(current_turf == target)
+			return
+		sleep(2)
+
+
```
Comment (DGamerL) [inline code/game/objects/effects/effect_system/effects_water.dm:43]:
```suggestion
```

File: code/game/objects/items/weapons/extinguisher.dm [hunk 4]
```diff
@@ -96,90 +96,74 @@
...
+			sleep(2)
+			step(C, movementdirection)
+			if(istype(C))
+				C.propelled = 0
+			sleep(3)
+			step(C, movementdirection)
+			sleep(3)
+			step(C, movementdirection)
+			sleep(3)
+			step(C, movementdirection)
```
Comment (Sirryan2002) [inline code/game/objects/items/weapons/extinguisher.dm:146]: make this into it's own proc and use `INVOKE_ASYNC()`

File: code/game/objects/items/weapons/extinguisher.dm [hunk 5]
```diff
@@ -96,90 +96,74 @@
...
+	if(!COOLDOWN_FINISHED(src, last_use))
+		return
+
+	COOLDOWN_START(src, last_use, 2 SECONDS)
+
+	if(reagents.chem_temp > 300 || reagents.chem_temp < 280)
+		add_attack_logs(user, target, "Sprayed with superheated or cooled fire extinguisher at Temperature [reagents.chem_temp]K")
+	playsound(loc, 'sound/effects/extinguish.ogg', 75, TRUE, -3)
+
+	var/direction = get_dir(src,target)
```
Comment (Sirryan2002) [inline code/game/objects/items/weapons/extinguisher.dm:114]:
```suggestion
	var/direction = get_dir(src, target)
```

File: code/game/objects/items/weapons/extinguisher.dm [hunk 6]
```diff
@@ -96,90 +96,74 @@
...
+	var/list/the_targets = list(T, T1, T2)
+	if(precision)
+		var/turf/T3 = get_step(T1, turn(direction, 90))
+		var/turf/T4 = get_step(T2, turn(direction, -90))
+		the_targets = list(T, T1, T2, T3, T4)
+
+	for(var/a in 1 to 5)
+		var/obj/effect/particle_effect/water/Water = new /obj/effect/particle_effect/water(get_turf(src))
+		Water.create_reagents(5)
+		reagents.trans_to(Water, 1)
```
Comment (Sirryan2002) [inline code/game/objects/items/weapons/extinguisher.dm:162]:
```suggestion
		var/obj/effect/particle_effect/water/water = new /obj/effect/particle_effect/water(get_turf(src))
		water.create_reagents(5)
		reagents.trans_to(water, 1)
```
snake_case vars

**Review by Sirryan2002 — APPROVED** [review]
code is fine, could use a bit of cleanup if you get around to it b4 merge

**Review by S34NW — APPROVED** [review]

**Pull request merged**